starport/pkg/cosmosclient/cosmosclient.go

```diff
@@ -17,12 +21,169 @@ import (
+	RPC *rpchttp.HTTP
+
+	// Factory is a Cosmos SDK tx factory.
+	Factory tx.Factory
+
+	// context is a Cosmos SDK client context.
+	Context client.Context
+
+	// Keyring is a Cosmos SDK keyring.
+	Keyring keyring.Keyring
+
+	nodeAddress        string
+	out                io.Writer
+	chainID            string
+	homePath           string
+	keyringServiceName string
+	keyringBackend     KeyringBackend
+}
+
+// Option configures your client.
+type Option func(*Client)
+
+// WithNodeAddress sets the node address of your chain. when this option is not provided
+// `http://localhost:26657` used as default.
+func WithNodeAddress(addr string) Option {
+	return func(c *Client) {
+		c.nodeAddress = addr
+	}
+}
+
+// WithHome sets the data dir of your chain. this option is used to access to your chain's
+// file based keyring which is only needed when you deal with creating and signing transactions.
+// when it is not provided, your data dir will be assumed as `$HOME/.your-chain-id`.
+func WithHome(path string) Option {
+	return func(c *Client) {
+		c.homePath = path
+	}
+}
+
+// WithKeyringServiceName used as the keyring's name when you are using OS keyring backend.
+// by default it is `cosmos`.
+func WithKeyringServiceName(name string) Option {
+	return func(c *Client) {
+		c.keyringServiceName = name
+	}
+}
+
+// WithKeyringBackend sets your keyring backend. by default it is `test`.
+func WithKeyringBackend(backend KeyringBackend) Option {
+	return func(c *Client) {
+		c.keyringBackend = backend
+	}
+}
+
+// New creates a new client with given options.
+func New(ctx context.Context, options ...Option) (Client, error) {
+	c := Client{
+		nodeAddress:        defaultNodeAddress,
+		out:                io.Discard,
+		keyringServiceName: sdktypes.KeyringServiceName(),
+		keyringBackend:     KeyringTest,
+	}
+
+	var err error
+
+	for _, apply := range options {
+		apply(&c)
+	}
+
+	if c.RPC, err = rpchttp.New(c.nodeAddress, "/websocket"); err != nil {
+		return Client{}, err
+	}
+
+	statusResp, err := c.RPC.Status(ctx)
+	if err != nil {
+		return Client{}, err
+	}
+
+	c.chainID = statusResp.NodeInfo.Network
+
+	if c.homePath == "" {
+		c.homePath = os.ExpandEnv(fmt.Sprintf("$HOME/.%s", c.chainID))
+	}
+
+	if c.Keyring, err = keyring.New(c.keyringServiceName, string(c.keyringBackend), c.homePath, os.Stdin); err != nil {
+		return Client{}, err
+	}
+
+	c.Context = newContext(c.RPC, c.out, c.chainID, c.homePath).WithKeyring(c.Keyring)
+	c.Factory = newFactory(c.Context)
+
+	return c, nil
+}
+
+// Address returns the account address from account name.
+func (c Client) Address(accountName string) (sdktypes.AccAddress, error) {
+	accountInfo, err := c.Keyring.Key(accountName)
+	if err != nil {
+		return nil, err
+	}
+	return accountInfo.GetAddress(), nil
+}
+
+// BroadcastTx creates and broadcasts a tx with given messages for account.
+func (c Client) BroadcastTx(accountName string, messages ...sdktypes.Msg) (*sdktypes.TxResponse, error) {
```

Comment:
I think we don't need the variadic function. In this case, we always pass an array for this method

```suggestion
func (c Client) BroadcastTx(accountName string, messages []sdktypes.Msg) (*sdktypes.TxResponse, error) {
```

Reply:
cosmosclient pkg is for public use, and it's valid to broadcast a tx with a single msg.